relayer/src/chain/cosmos.rs

```diff
@@ -839,28 +840,37 @@ fn packet_query(request: &QueryPacketEventDataRequest, seq: &Sequence) -> Result
     ))
 }
 
-// Extract the packet events from the query_tx RPC response. The response includes the full set of events
-// from the Tx-es where there is at least one request query match.
-// For example, the query request asks for the Tx for packet with sequence 3, and both 3 and 4 were
-// committed in one Tx. In this case the response includes the events for 3 and 4.
+// Extract the packet events from the query_txs RPC response. For any given
+// packet query, there is at most one Tx matching such query. Moreover, a Tx may
+// contain several events, but a single one must match the packet query.
+// For example, if we're querying for the packet with sequence 3 and this packet
+// was committed in some Tx along with the packet with sequence 4, the response
+// will include both packets. For this reason, we iterate all packets in the Tx,
+// searching for those that match (which must be a single one).
 fn packet_from_tx_search_response(
     request: &QueryPacketEventDataRequest,
     seq: Sequence,
-    response: &tendermint_rpc::endpoint::tx_search::Response,
+    mut response: tendermint_rpc::endpoint::tx_search::Response,
 ) -> Result<Option<IBCEvent>, Error> {
-    // TODO: remove loop as `response.txs.len() <= 1`
-    for r in response.txs.iter() {
+    assert!(
+        response.txs.len() <= 1,
+        "packet_from_tx_search_response: unexpected number of txs"
+    );
+    if let Some(r) = response.txs.pop() {
         let height = r.height;
         if height.value() > request.height.revision_height {
-            continue;
+            return Ok(None);
         }
 
-        for e in r.tx_result.events.iter() {
-            if e.type_str != request.event_id.as_str() {
-                continue;
-            }
+        let mut matching = Vec::new();
+        for e in r.tx_result.events {
+            assert_eq!(
+                e.type_str,
+                request.event_id.as_str(),
+                "packet_from_tx_search_response: unexpected event type"
+            );
```

Comment:
This doesn't seem right. It is true that there is one single transaction but it may include multiple events. I think what was there before was correct:
```suggestion
            if e.type_str != request.event_id.as_str() {
                continue;
            }
```

Reply:
Thanks! This was the reason why the tests were failing. Fixed in b0a9aa0a36126cbeec414f5ac659b997cf594594.